src/agents/bee/runners/default/runner.ts

```diff
@@ -443,16 +221,178 @@ export class BeeAgentRunner extends Serializable {
     }).get();
   }
 
-  createSnapshot() {
-    return {
-      input: shallowCopy(this.input),
-      options: shallowCopy(this.options),
-      memory: this.memory,
-      failedAttemptsCounter: this.failedAttemptsCounter,
+  @Cache({ enumerable: false })
+  protected get renderers() {
+    const self = {
+      user: {
+        message: ({ prompt }: BeeRunInput) =>
+          prompt !== null || this.input.memory.isEmpty()
+            ? BaseMessage.of({
+                role: Role.USER,
+                text: prompt ?? "",
+                meta: {
+                  createdAt: new Date(),
+                },
+              })
+            : undefined,
+      },
+      system: {
+        variables: {
+          tools: async () => {
+            return await Promise.all(
+              this.input.tools.map(async (tool) => ({
+                name: tool.name,
+                description: tool.description.replaceAll("\n", ".").replace(/\.$/, "").concat("."),
+                schema: JSON.stringify(
+                  await tool.getInputJsonSchema(),
+                  (() => {
+                    const ignoredKeys = new Set(["minLength", "maxLength", "$schema"]);
```

Comment:
Nitpick: This works recursively right? It's not equivalent to the `R.omit` that granite runner used before, but it's probably better this way.

Reply:
Yes, this works recursively. Hence, it is not equivalent to the original implementation. However, it is a fix of the original implementation because it never really worked (the schema object typically doesn't have such keywords at a top level (maybe just `$schema`)).